compiler/vm/src/mir_to_lir.rs

```diff
@@ -157,9 +157,24 @@ impl<'c> LoweringContext<'c> {
                     self.emit(id, Instruction::PushFromStack(offset));
                 }
             }
-            Expression::Symbol(symbol) => {
-                let tag = Tag::create_from_str(&mut self.lir.constant_heap, symbol, None);
-                self.constants.insert(id, tag.into());
+            Expression::Tag { symbol, value } => {
+                let symbol = Text::create(&mut self.lir.constant_heap, symbol);
+
+                match value {
+                    Some(value) => {
+                        if let Some(value) = self.constants.get(value) {
+                            let tag = Tag::create(&mut self.lir.constant_heap, symbol, *value);
+                            self.constants.insert(id, tag.into());
+                        } else {
+                            self.emit_reference_to(*value);
+                            self.emit(id, Instruction::CreateTag { symbol });
+                        }
+                    }
+                    None => {
+                        let tag = Tag::create(&mut self.lir.constant_heap, symbol, None);
+                        self.constants.insert(id, tag.into());
+                    }
+                }
```

Comment:
Maybe the outer `match` on an `Option` could also be converted to an `if let` for consistency